Issue #688 - Reorganise `validate_forecast()`

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #688. 
This PR closes #689.  
This PR closes #773.

In #688 and in particular in #753 we discussed the relationship between `validate_forecast()` and `as_forecast()` and decided on the following: 
- `as_forecast()` converts from a non-`forecast` object to a `forecast` object, validates it and returns the object
- `assert_forecast()` re-validates an existing `forecast` object and returns `invisble(NULL)`
- `validate_forecast()` is like `assert_forecast()`, but returns the object so it can be used in a pipe. 

This PR
- renames the current `validate_forecast()` to `assert_forecast()` and changed it to return `invisble(NULL)`
- introduces a `silent` argument to `validate_forecast()` (that effectively prevents one single message from being thrown)
- uses that new `silent` argument, e.g. when printing
- creates a new `validate_forecast()` function that calls `assert_forecast()` and returns the data
- has a change with a questionable relationship to this PR! <3
  - namely it renames `validate_general()` to `assert_forecast_generic()`
- updates the News file
- updates the manuscript
- does not have any new tests. Should it? I had a vague feeling it should, but I'm also a bit unclear what to test....

Further question: 
Should `as_forecast()` get a `silent` argument as well? We have quite a large number of `suppressMessages()` calls in our tests, which maybe is a hint...

[Describe the changes that you made in this pull request.]

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

seabbs commented on NAMESPACE:
```diff
@@ -18,15 +23,12 @@ S3method(score,forecast_binary)
 S3method(score,forecast_quantile)
 S3method(score,forecast_sample)
-S3method(validate_forecast,default)
-S3method(validate_forecast,forecast_binary)
```
does validate_forecast need to be a s3 method now or can it just depend on assert_forecast dispatching underneath?

seabbs commented on R/forecast.R:
```diff
@@ -253,29 +265,45 @@ validate_forecast.forecast_point <- function(forecast, forecast_type = NULL, ...
+#' @inherit assert_forecast params return examples
+#' @export
+#' @keywords check-forecasts
+validate_forecast <- function(forecast, forecast_type = NULL, silent = FALSE) {
```
Is your namespace up to date? This isn't showing up is it?

nikosbosse replied:
I think it's not showing up because there was previously a different `validate_forecast()` function

seabbs commented on R/forecast.R:
```diff
@@ -318,8 +347,8 @@ validate_general <- function(data) {
-  if (!is.logical(number_quantiles_samples)) {
-    warning(number_quantiles_samples)
+  if (!is.logical(number_quantiles_samples) && !silent) {
+    cli_warn(number_quantiles_samples)
```
missing roxgyen importFrom statement

seabbs replied:
This looks like a new test gap

seabbs commented on R/forecast.R:
```diff
@@ -292,13 +320,14 @@ validate_forecast.forecast_sample <- function(forecast, forecast_type = NULL, ..
 #' @export
 #' @keywords internal_input_check
-validate_general <- function(data) {
+assert_forecast_generic <- function(data, silent = FALSE) {
```
its not really silent as it will still flag errors maybe verbose is a better way to put it.

seabbs reviewed (changes requested):
I think I really like this. A few questions/comments

seabbs commented:
> Should as_forecast() get a silent argument as well? We have quite a large number of suppressMessages() calls in our tests, which maybe is a hint...

I think if we rename it to be `verbose` then yes

nikosbosse commented:
Latest update: 
- added missing importFrom for `cli_warn()` 
- changed `silent` to `verbose`
- made sure those `verbose` statements actually get passed on to `assert_forecast()`
- added new unit tests, in particular making sure that we get a warning when there are different numbers of quantiles, but no warning when printing (as `assert_forecast()` is in silent mode)
- updated docs/namespace

seabbs reviewed (approved):
Nice.
